source/elements/oneDAL/source/algorithms/nearest_neighbors/knn_classification.rst

```diff
@@ -1,3 +1,198 @@
+Training
+--------
+Given the training set :math:`X = \{ x_1, \ldots, x_n \}` of
+:math:`p`-dimensional feature vectors, the vector of class labels :math:`y` and
+a positive integer :math:`k`, where :math:`y_i \in \{ 0, \ldots, c-1 \}`,
+:math:`1 \leq i \leq n`, :math:`c` is the number of classes and :math:`k` is
+number of nearest neighbors specified by the user. The problem is to build a
+model that allows efficient distance computation between the feature vectors in
+:math:`X` and :math:`X'` at the inference stage.
+
+
+.. _t_math_brute_force:
+
+Training method: *brute-force*
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The training operation produces the model that simply stores all the
+feature vectors from the initial training set :math:`X`.
+
+
+.. _t_math_kd_tree:
+
+Training method: *k-d tree*
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The training operation builds a :math:`k`-:math:`d` tree (for more details, see
+:ref:`kd_tree`) that partitions the training set :math:`X`.
+
+
+.. _i_math:
+
+Inference
+---------
+Given the set :math:`X' = \{ x_1', \ldots, x_m' \}` of :math:`p`-dimensional
+feature vectors, the user defined number of nearest neighbors :math:`k` and the
+model produced by the training operation. The inference operation predicts the
+label for each :math:`x_i'`, :math:`1 \leq i \leq n`, by performing the
+following steps:
+
+#. Identify the set :math:`N(x_i') \subseteq X` of the :math:`k` feature vectors
+   in the training set that are nearest to :math:`x_i'` with respect to the
+   Euclidean distance.
+
+#. Estimate the conditional probability for the :math:`l`-th class as the
+   fraction of vectors in :math:`N(x_i')` whose labels :math:`y_i` are equal to
+   :math:`l`:
+
+   .. math::
+      :label: p_predict
+
+      P(y_i' = l \; | \; x_i') = \frac{1}{| N(x_i') |}
+      \Big| \big\{ x_r \in N(x_i') : y_r = l \big\} \Big|,
+      \quad 1 \leq i \leq m, \; 0 \leq l < c.
+
+#. Use the class with the largest probability as prediction for the feature
+   vector :math:`x_i'`:
+
+   .. math::
+      :label: y_predict
+
+      y_i' = \mathrm{arg}\max_{0 \leq l < c} P(y_i' = l \; | \; x_i'),
+      \quad 1 \leq i \leq m.
+
+
+.. _i_math_brute_force:
+
+Inference method: *brute-force*
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The inference operation determines the set :math:`N(x_j')` of the nearest
+feature vectors by iterating over all the pairs :math:`(x_j', x_i)` in the
+implementation defined order, :math:`1 \leq i \leq n`, :math:`1 \leq j \leq m`.
+The final prediction is computed according to the equations :eq:`p_predict` and
+:eq:`y_predict`.
+
+
+.. _i_math_kd_tree:
+
+Inference method: *k-d tree*
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
+
+---------------------
+Programming Interface
+---------------------
+
+Computational methods
+---------------------
+.. namespace:: onedal::classification::knn::method
+
+::
+
+   namespace method {
+      struct bruteforce {};
+      struct kd_tree {};
+      using by_default = brutforce;
+   } // namespace method
```

Comment:
does this mean that with `dalapi` the namespaces are not taken from the header files directly, but managed in rst files and require a bit of copy-pasting? is it a limitation or not? I'm trying to figure out why you are not dealing with it in the same way as with everything else.

Reply:
There is no any limitation, it's matter of time. I'm going to add namespaces and typedef/using support soon, it's a bit harder to implement than for classes.